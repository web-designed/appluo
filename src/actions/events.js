import uuid from 'uuid'
import moment from 'moment'
import database from '../firebase/firebase'


//*******************************************************
// EVENTS ACTION GENERATORS
//*******************************************************

export const addEvent = (event) => ({
   type:'ADD_EVENT',
   event
})

export const startAddEvent = (eventData = {}) => {
   return (dispatch) => {
      const {
         note = '', 
         cleaner = 'unknown', 
         cleanedAt = 0,
         createdAt = 0,
         place = ''
      } = eventData

      const event = { note, cleaner, cleanedAt, createdAt, place }

      return database.ref('events').push(event).then((ref) => { // add return to chain promises in the tests
         dispatch(addEvent({
            id: ref.key,
            ...event
         }))
      })
   }
}


export const removeEvent = (id) => ({
   type: 'REMOVE_EVENT',
   id
})

export const editEvent = (id, update) => ({
   type: 'EDIT_EVENT', 
   id,
   update
})

export const addComment = (eventId, { commentBody, commenter }) => ({
   type: 'ADD_COMMENT',
   eventId,
   comment: {
      id: uuid(),
      createdAt: moment().valueOf(),
      commenter,
      commentBody
   }
})

export const removeComment = (eventId, commentId) => ({
   type:'REMOVE_COMMENT',
   eventId,
   commentId
})

export const editComment = (eventId, commentId, update) => ({
   type:'EDIT_COMMENT',
   eventId,
   commentId,
   update
})