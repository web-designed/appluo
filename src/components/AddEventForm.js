import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { Link } from 'react-router-dom'

export class AddEventForm extends React.Component {

   state = {
      date: this.props.cleanedAt ? moment(this.props.cleanedAt) : moment(),
      focused: false,
      note: this.props.note ? this.props.note : '',
      place: this.props.place ? this.props.place : this.props.settings.places[0],
      error: '',
      cleaner: this.props.cleaner ? this.props.cleaner : this.props.settings.cleaners[0],
   }

   onCleanerChange = (e) => {
      const cleaner = e.target.value
      this.setState(() => ({ cleaner }))
   }

   onPlaceChange = (e) => {
      const place = e.target.value
      this.setState(() => ({ place }))
   }

   onNoteChange = (e) => {
      const note = e.target.value
      this.setState(() => ({ note }))
   }

   onDateChange = (date) => {
      this.setState(() => ({ date }))
   }

   onFocusChange = ({focused}) => {
      this.setState(() => ({focused}))
   }

   handleSubmit = (e) => {

      e.preventDefault()
      
      const event = {
         cleanedAt: this.state.date.valueOf(),
         createdAt: moment().valueOf(),
         cleaner: this.state.cleaner,
         place: this.state.place,
         note: this.state.note,
      }

      this.props.handleSubmit(event)
   }

   render(){

      console.log(this.props)

      return(
         <div>
            <form onSubmit={this.handleSubmit}>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="cleaner">Name</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="cleaner"
                     name="cleaner" 
                     onChange={this.onCleanerChange}
                     value={this.state.cleaner.toLowerCase()}
                  >
                     {
                        this.props.settings.cleaners.map( cleaner => {
                           return <option value={cleaner.toLowerCase()} key={cleaner.toLowerCase()}>{cleaner}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="place">Place</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="place"
                     name="place" 
                     onChange={this.onPlaceChange}
                     value={this.state.place.toLowerCase()}
                  >
                     {
                        this.props.settings.places.map( place => {
                           return <option value={place.toLowerCase()} key={place}>{place}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="date">Date</label>
                  </div>
                  <SingleDatePicker 
                     date={this.state.date}
                     onDateChange={this.onDateChange}
                     focused={this.state.focused}
                     onFocusChange={this.onFocusChange}
                     numberOfMonths={1}
                     isOutsideRange={() => false}
                  />
               </div>
               <div class="form-group">
                  <textarea 
                     class="form-control"
                     value={this.state.note}
                     name="note" 
                     placeholder="Add your note"
                     onChange={this.onNoteChange}
                  ></textarea>
               </div>
               <div class="form-group text-right">
                  <Link class="text-danger mr-2" to={'/'}>Cancel</Link>
                  <button class="btn btn-primary">Submit</button>
               </div>
            </form>
         </div>
      )
   }
}

const mapStateToProps = ({ settings }) => ({
   settings: settings
})

export default connect(mapStateToProps)(AddEventForm)