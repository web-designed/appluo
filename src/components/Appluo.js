import React from 'react'
import EventsList from './EventsList'

class Appluo extends React.Component {
   render() {
      return (
         <div>
            <EventsList />
         </div>
      )
   }
}

export default Appluo