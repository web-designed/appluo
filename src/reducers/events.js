//*******************************************************
// EVENTS REDUCERS
//*******************************************************

   const eventsReducerDefaultState = []
   const eventsReducer = (state = eventsReducerDefaultState, action) => {
      switch(action.type){
         case 'ADD_EVENT':
            return state.concat(action.event)
         case 'REMOVE_EVENT':
            return state.filter(({ id }) => id !== action.id)
         case 'EDIT_EVENT':
            return state.map((event) => {
               if(event.id === action.id) {
                  return {
                     ...event,
                     ...action.update
                  }
               } else {
                  return event
               }
            })
         case 'ADD_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: [...event.comments, action.comment]
                  }
               } else {
                  return event
               }
            })
         case 'REMOVE_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.filter((comment) => {
                        return comment.id !== action.commentId
                     })
                  }
               } else {
                  return event
               }
            })
         case 'EDIT_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.map((comment) => {
                        if(comment.id === action.commentId){
                           return {
                              ...comment,
                              ...action.update
                           }
                        } else {
                           return comment
                        }
                     }) 
                  }
               } else {
                  return event
               }
            })
         default:
            return state
      }
   }

export default eventsReducer