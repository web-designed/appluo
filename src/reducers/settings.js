import moment from 'moment'

//*******************************************************
// SETTINGS REDUCER
//*******************************************************

   const settingsReducerDefaultState = {
      wgName: 'The Cleaning Team',
      slogan: 'Manage your cleaning plan',
      places: ['Kitchen', 'Bathroom', 'Saloon'],
      cleaners: ['John', 'Alex', 'Tom', 'Steve'],
      createdAt: moment('2018-12-01'),
      notifications: {
         event: {
            afterDelete: 'The cleaning was successfully removed',
            notFound: "The cleaning doesn't exist or has been removed"
         },
         comment: {
            validate: 'Please add your comment'
         }
      }
   }
   const settingsReducer = (state = settingsReducerDefaultState, action) => {
      switch(action.type){
         default:
            return state
      }
   } 

   export default settingsReducer