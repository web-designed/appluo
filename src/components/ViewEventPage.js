import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import CommentsList from './CommentsList'
import moment from 'moment'
import { startRemoveEvent } from '../actions/events'
import CommentAddForm from './CommentAddForm'

export class ViewEventPage extends React.Component {

   state = {
      eventDeletedNotification: ''
   }

   handleRemoveEvent = () => {
      this.setState(() => ({
         eventDeletedNotification: this.props.notifications.afterDelete
      }))
      setTimeout(() => {
         this.props.startRemoveEvent(this.props.event.id)
         this.props.history.push('/')
      }, 2000);
   }

   render(){
      if(this.props.event){
         return (
            <div class="container-fluid">
               {
                  !this.state.eventDeletedNotification ? (
                     <div class="pt-3 row justify-content-center">
                        <div class="col-md-6">
                           <div class="card mb-3">
                              <div class="card-header d-flex justify-content-between align-items-center">
                                 <div>
                                    <h5 class="mb-0">{this.props.event.place}</h5>
                                 </div>
                                 <div>
                                    <small class="text-secondary">cleaned by:</small> {this.props.event.cleaner} <small class="text-secondary">on:</small> {moment(this.props.event.cleanedAt).format('YYYY/MM/D')}
                                 </div>
                              </div>
                              {
                                 this.props.event.note && (
                                    <div class="card-body">
                                       <h6 class="card-subtitle mb-2 text-muted"></h6>
                                       <p class="card-text">{this.props.event.note}</p>
                                    </div>
                                 )
                              }
                              <div class="card-footer text-muted">
                                 <button class="btn btn-danger mr-1" onClick={this.handleRemoveEvent}>remove</button> 
                                 <Link class="btn btn-primary" to={`/edit/${this.props.event.id}`}>Edit</Link>
                              </div>
                           </div>
                           <div>
                              <CommentsList currentEvent={this.props.event} />
                              <CommentAddForm currentEventId={this.props.event.id} />
                           </div>
                        </div>
                     </div>
                  ) : (
                     <div class="pt-3 row justify-content-center">
                        <div class="col-md-6">
                           <h3>{this.state.eventDeletedNotification}</h3>
                           <Link to="/"><button class="btn btn-primary">Dashboard</button></Link>
                        </div>
                     </div>
                  )
               }
            </div>
         )
      } else {
         return (
            <div class="container-fluid">
               <h1>{this.props.notifications.notFound}</h1>
               <Link to="/"><button>Dashboard</button></Link>
            </div>
         )
      }
   }
}

const mapDispatchToProps = (dispatch, props) => ({
   startRemoveEvent: (eventId) => { dispatch(startRemoveEvent(eventId)) }
})

const mapStateToProps = (state, props) => ({
   notifications: state.settings.notifications.event,
   event: state.events.find((event) => {
      return event.id === props.match.params.id
   })
})

export default connect(mapStateToProps, mapDispatchToProps)(ViewEventPage)