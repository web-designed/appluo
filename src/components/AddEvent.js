import React from 'react'
import AddEventForm from './AddEventForm'

const AddEvent = () => {
   return(
      <div>
         <h1>Add a cleaning event</h1>
         <AddEventForm />
      </div>
   )
}

export default AddEvent