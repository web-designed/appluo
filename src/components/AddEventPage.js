import React from 'react'
import AddEventForm from './AddEventForm'
import { addEvent } from '../actions/events'
import { connect } from 'react-redux'

const AddEventPage = (props) => (
   <div>
      <h1>Add a cleaning</h1>
      <AddEventForm 
         handleSubmit={ (event) => {
            props.onSubmit(event)
            props.history.push('/')
         }} 
         history={props.history} 
      />
   </div>
)

const mapDispatchToProps = (dispatch) => ({
   onSubmit: (expense) => dispatch(addEvent(expense))
})

export default connect(undefined, mapDispatchToProps)(AddEventPage)