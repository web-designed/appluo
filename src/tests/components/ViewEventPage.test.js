import React from 'react'
import { ViewEventPage } from '../../components/ViewEventPage'
import { shallow } from 'enzyme'
import events from '../fixtures/events'
import settings from '../fixtures/settings'

let wrapper, removeEvent, history

beforeEach(() => {
   removeEvent = jest.fn()
   history = {
      push: jest.fn()
   }
   wrapper = shallow(
      <ViewEventPage
         notifications={settings.notifications.event} 
         removeEvent={removeEvent} 
         event={events[0]}
         history={history}
      />
   )
})

test('should render the ViewEventPage correctly if no event found', () => {
   const wrapperEmpty = shallow(<ViewEventPage  notifications={settings.notifications.event}  />)
   expect(wrapperEmpty).toMatchSnapshot()
})

test('should render the ViewEventPage correctly if event found', () => {
   expect(wrapper).toMatchSnapshot()
})

test('should handle the handleRemoveEvent', () => {
   wrapper.find('button').at(0).simulate('click')
   expect(wrapper.state('eventDeletedNotification')).toBe(settings.notifications.event.afterDelete)
   setTimeout(() => {
      expect(removeEvent).toHaveBeenLastCalledWith(events[0].id)
      expect(history.push).toHaveBeenLastCalledWith('/')
   }, 2000);
})