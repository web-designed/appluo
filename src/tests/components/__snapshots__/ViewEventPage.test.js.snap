// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the ViewEventPage correctly if event found 1`] = `
<div
  class="container-fluid"
>
  <div
    class="pt-3 row justify-content-center"
  >
    <div
      class="col-md-6"
    >
      <div
        class="card mb-3"
      >
        <div
          class="card-header d-flex justify-content-between align-items-center"
        >
          <div>
            <h5
              class="mb-0"
            >
              Bad
            </h5>
          </div>
          <div>
            <small
              class="text-secondary"
            >
              cleaned by:
            </small>
             
            Kylu
             
            <small
              class="text-secondary"
            >
              on:
            </small>
             
            1970/01/1
          </div>
        </div>
        <div
          class="card-body"
        >
          <h6
            class="card-subtitle mb-2 text-muted"
          />
          <p
            class="card-text"
          >
            Alles sauber gemacht
          </p>
        </div>
        <div
          class="card-footer text-muted"
        >
          <button
            class="btn btn-danger mr-1"
            onClick={[Function]}
          >
            remove
          </button>
          <Link
            class="btn btn-primary"
            replace={false}
            to="/edit/0"
          >
            Edit
          </Link>
        </div>
      </div>
      <div>
        <CommentsList
          currentEvent={
            Object {
              "cleanedAt": 0,
              "cleaner": "Kylu",
              "comments": Array [
                Object {
                  "commentBody": "this is my comment",
                  "commenter": "Kylu",
                  "createdAt": 125,
                  "id": "1",
                },
                Object {
                  "commentBody": "this is my another comment",
                  "commenter": "Chris",
                  "createdAt": 130,
                  "id": "2",
                },
              ],
              "createdAt": 0,
              "id": "0",
              "note": "Alles sauber gemacht",
              "place": "Bad",
            }
          }
        />
        <Connect(CommentAddForm)
          currentEventId="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the ViewEventPage correctly if no event found 1`] = `
<div
  class="container-fluid"
>
  <h1>
    The cleaning doesn't exist or has been removed
  </h1>
  <Link
    replace={false}
    to="/"
  >
    <button>
      Dashboard
    </button>
  </Link>
</div>
`;
